scylla/src/frame/value.rs

```diff
@@ -104,27 +108,62 @@ impl SerializedValues {
         SerializedValues {
             serialized_values: Vec::new(),
             values_num: 0,
+            contains_names: false,
         }
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
         SerializedValues {
             serialized_values: Vec::with_capacity(capacity),
             values_num: 0,
+            contains_names: false,
         }
     }
 
+    pub fn with_names(&self) -> bool {
+        self.contains_names
+    }
+
     /// A const empty instance, useful for taking references
     pub const EMPTY: &'static SerializedValues = &SerializedValues::new();
 
     /// Serializes value and appends it to the list
     pub fn add_value(&mut self, val: &impl Value) -> Result<(), SerializeValuesError> {
+        if self.contains_names {
+            return Err(SerializeValuesError::MixingNamedAndNotNamedValues);
+        }
+        if self.values_num == i16::max_value() {
+            return Err(SerializeValuesError::TooManyValues);
+        }
+
+        let len_before_serialize: usize = self.serialized_values.len();
+
+        if let Err(e) = val.serialize(&mut self.serialized_values) {
+            self.serialized_values.resize(len_before_serialize, 0);
+            return Err(SerializeValuesError::from(e));
+        }
+
+        self.values_num += 1;
+        Ok(())
+    }
+
+    pub fn add_named_value(
+        &mut self,
+        name: &str,
+        val: &impl Value,
+    ) -> Result<(), SerializeValuesError> {
+        if self.values_num > 0 && !self.contains_names {
+            return Err(SerializeValuesError::MixingNamedAndNotNamedValues);
+        }
+        self.contains_names = true;
         if self.values_num == i16::max_value() {
             return Err(SerializeValuesError::TooManyValues);
         }
 
         let len_before_serialize: usize = self.serialized_values.len();
 
+        types::write_string(name, &mut self.serialized_values).unwrap();
```

Comment:
My first idea was indeed to get rid of the unwrap, but I decided to leave it after I realized that it could only happen when a programmer provides a wrong or not validated buffer as SerializedValues. But maybe you're right that panicking is too harsh and we should just prefer a regular error in here anyway.